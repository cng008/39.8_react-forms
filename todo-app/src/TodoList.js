import React, { useState } from 'react'
import Todo from './Todo'
import NewTodoForm from './NewTodoForm'

/** render the NewTodoForm component
 * render the list of Todo components
 */

const TodoList = () => {
  const [todos, setTodos] = useState([])

  // add a new todo

  const addTodo = newTodo => {
    setTodos(todos => [...todos, newTodo])
  }

  // update a todo with updatedTask
  const update = (id, updatedTask) => {
    setTodos(todos =>
      todos.map(todo =>
        todo.id === id ? { ...todo, task: updatedTask } : todo
      )
    )
  }

  // delete a todo by id
  const remove = id => {
    setTodos(todos => todos.filter(todo => todo.id !== id))
  }
  return (
    <div className="ToDoList">
      <h1>Todo App</h1>
      <NewTodoForm addTodo={addTodo} />
      <ul>
        {todos.map(todo => (
          <Todo
            key={todo.id}
            id={todo.id}
            task={todo.task}
            handleRemove={remove}
            update={update}
          />
        ))}
      </ul>
    </div>
  )
}

export default TodoList
