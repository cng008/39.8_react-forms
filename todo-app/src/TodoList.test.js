import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import TodoList from './TodoList'
import '@testing-library/jest-dom/extend-expect'

function addTodo(todoList) {
  const taskInput = todoList.getByLabelText('Task:')
  fireEvent.change(taskInput, { target: { value: 'write tests' } })
  const submitButton = todoList.getByText('+')
  fireEvent.click(submitButton)
}

describe('<TodoList /> rendering', () => {
  it('renders without crashing', function () {
    render(<TodoList />)
  })

  it('matches snapshot', function () {
    const { asFragment } = render(<TodoList />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('can add a todo', function () {
    const list = render(<TodoList />)
    addTodo(list)

    // expect form to clear and todo to be on the page
    expect(list.getByLabelText('Task:')).toHaveValue('')
    expect(list.getByText('write tests')).toBeInTheDocument()
    expect(list.getByTestId('Edit')).toBeInTheDocument()
    expect(list.getByTestId('Remove')).toBeInTheDocument()
  })
})
