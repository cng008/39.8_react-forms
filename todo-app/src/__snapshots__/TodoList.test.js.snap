// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> rendering matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ToDoList"
  >
    <h1>
      Todo App
    </h1>
    <form
      class="NewTodoForm"
    >
      <label
        class="NewTodoForm-Label"
        for="task"
      >
        Task:
      </label>
      <input
        class="NewTodoForm-Input"
        id="task"
        name="task"
        placeholder="Enter a task"
        type="text"
        value=""
      />
      <button>
        +
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
